Required changes to support Search App

imalberto opened the pull request:
CHANGELOG
- added ac.url.find() API to support mojito-\* related dependencies
- added ac.url.linkTo() helper to support url generation based on route name
- lib/mojito.js now integrates with expressjs via "Branding"
- lib/middleware : app specific middleware are now passed a `midConfig` object during registration
- lib/router : uses `getStaticAppConfig().routesFiles` as the source to load routes configuration

TODO
- [x] fix unit tests
- [x] fix functional tests
- [x] update examples
- [ ] update `next` docs

Commit:
Merge pull request #1309 from lzhan/next

Merge pull requests to next branch

Commit:
Merge pull request #1311 from lzhan/next

rm wait() from func tests

Commit:
Merge pull request #1315 from lzhan/next

Add this fix to next branch: Fix issue #1306: ac.partial.render() throws error for handlebar custom helper

imalberto committed:
- lib/dispatcher : removed unused function
- lib/app/autoload/route-maker : added support to look up and resolve path based on a route name

imalberto committed:
- lib/dispatcher : removed unused function
- lib/app/autoload/route-maker : added support to look up and resolve path based on a route name

imalberto committed:
Refactored lib/mojito to use "Branding" as an extension to Express app.

imalberto committed:
Refactored lib/mojito to use "Branding" as an extension to Express app.

imalberto committed:
First pass at adding support for ac.url.find()

imalberto committed:
First pass at adding support for ac.url.find()

imalberto commented:
/cc @caridy @drewfish

imalberto commented on lib/dispatcher.js:
```diff
@@ -64,13 +64,6 @@ module.exports = {
         }
 
-        // now looking at the last callback with the appropriate "annotations"
-        function extractAnnotations(route) {
```
Unused function from previous revisions. Removed in this PR.

imalberto commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -97,22 +97,126 @@ YUI.add('mojito-route-maker', function (Y, NAME) {
+                copy.params[key.name] = matches[j];
+            }
+
+            // Lookup the `call` by aliases
```
Not too happy about this block about how the `call` is decoded.

imalberto commented on lib/mojito.js:
```diff
@@ -58,6 +48,15 @@ var debug = require('debug')('mojito'),
 //  Mojito Global
 //  ----------------------------------------------------------------------------
 
+function extend(app, options) {
```
With this change, now `app.js` can do something like this:

```
var express = require('express'),
      libmojito = require('mojito'),
      app = express();

libmojito.extend(app, {
    context: {
        runtime: 'server',
        environment: 'staging',
        colo: 'sp1'
    }
});

// proceed as usual
app.use(libmojito.middleware()); 
... ....
```

caridy replied:
why is the `port` needed?

imalberto replied:
`port` is not needed, since the actual `listen` is done at the `app.js` level. Good point. Will be removing references to `port` in next commit.

imalberto committed:
- lib/middleware : BC change for applications that specify mojito-* middleware configurations via application.json (e.g. mojito-debug)
- lib/router: Fix to use static `appConfig.routesFiles` to load routes  configuration (mojito-debug uses RS addon to modify `routesFiles`)

imalberto committed:
- lib/middleware : BC change for applications that specify mojito-* middleware configurations via application.json (e.g. mojito-debug)
- lib/router: Fix to use static `appConfig.routesFiles` to load routes  configuration (mojito-debug uses RS addon to modify `routesFiles`)

imalberto committed:
Removed references to `port`.

imalberto committed:
Removed references to `port`.

imalberto committed:
Removed explicit reference to `express`.

imalberto committed:
Removed explicit reference to `express`.

imalberto committed:
Added unit tests for route-maker#find()

imalberto committed:
Added unit tests for route-maker#find()

imalberto committed:
Removed unused logging statement

imalberto committed:
Removed unused logging statement

imalberto committed:
- Added more lib/router tests
- Fixed broken unit tests

imalberto committed:
- Added more lib/router tests
- Fixed broken unit tests

Commit:
Merge pull request #1316 from lzhan/next

add more checking/waiting to the unstable tests

drewfish commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -97,22 +97,124 @@ YUI.add('mojito-route-maker', function (Y, NAME) {
+            len = copy.keys.length;
+            for (i = 0, j = 1; i < len; i += 1, j += 1) {
+                key = copy.keys[i];
+                copy.params[key.name] = matches[j];
```
If `j` is just one more than `i` it might make more sense to do `= matches[i + 1]` instead.

Commit:
update tests to make time/date related tests run fine on both grid and local machines

imalberto committed:
More unit tests for lib/router and lib/mojito.

imalberto committed:
More unit tests for lib/router and lib/mojito.

Commit:
update arrow version to 0.3.1

imalberto committed:
Updated all the examples/ `app.js` that functional tests uses.

imalberto committed:
Updated all the examples/ `app.js` that functional tests uses.

imalberto committed:
More fixes to `app.js` under examples/.

imalberto committed:
More fixes to `app.js` under examples/.

imalberto committed:
Functional tests fixes.

imalberto committed:
Functional tests fixes.

imalberto commented:
@caridy PR is ready to merge. The unit and func tests are OK when running individually , but failing as a whole when running of the entire suite, related to the issues @lzhan is investigating using the Locator feature from Arrow.

Once this PR is merged, @zhouyaoji can then merge his PRs related to examples and docs (since they've changed). See #1324 #1327

imalberto committed:
Removed use of Arrow locator since its causing timeout failures

imalberto committed:
Merged with origin/search-next

caridy commented on examples/developer-guide/adding_view_engines/app.js:
```diff
@@ -15,10 +15,12 @@ var debug = require('debug')('app'),
+mojito.extend(app);
 
-app.use(mojito.middleware());
+app.use(app.mojito.middleware());
```
middleware are monadic functions that has nothing to do with the state of the app. I wonder why you decided to move this from `mojito.middlware` to `app.mojito.middleware()`?

caridy commented on examples/developer-guide/adding_view_engines/app.js:
```diff
@@ -15,10 +15,12 @@ var debug = require('debug')('app'),
+app.use(app.mojito.middleware());
 app.mojito.attachRoutes();
-app.post('/tunnel', mojito.tunnelMiddleware());
+app.post('/tunnel', app.mojito.tunnelMiddleware());
```
same as above.

caridy commented on examples/developer-guide/dashboard/07_binders/app.js:
```diff
@@ -11,10 +11,14 @@
 app = express();
+app.set('port', process.env.PORT || 8666);
+libmojito.extend(app);
+mojito = app.mojito;
```
I don't like this.

caridy commented on lib/app/addons/ac/url.common.js:
```diff
@@ -30,6 +30,26 @@ YUI.add('mojito-url-addon', function(Y, NAME) {
+        @param {String} url the URL to find a route for.
+        @param {String} the HTTP method. Default value is "get".
+        **/
+        find: function(url, verb) {
```
I thought we were going to move this into the corresponding pkg that requires it.

imalberto replied:
sync'ed with @caridy offline. For now, the `find()` api will remain in the `ac.url` addon.

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -92,27 +60,131 @@ YUI.add('mojito-route-maker', function (Y, NAME) {
+        @param {String} verb the HTTP method.
+        @return {Object} the matched route object, or null
+        **/
+        find: function(uri, verb) {
```
same as above.

caridy commented on lib/app/autoload/route-maker.common.js:
```diff
@@ -92,27 +60,131 @@ YUI.add('mojito-route-maker', function (Y, NAME) {
-            // TODO How can we leverage express-map#pathTo ?
-            // return (CACHE.routes[name] && CACHE.routes[name].path) || null;
-            return null;
+        linkTo: function(name, context) {
```
why is the context needed here? is this the mojito context?

imalberto replied:
I think `params` is perhaps better suited here, since it will be used to resolved any parametrized urls.

caridy replied:
so, it is not the context of mojito, just a context object. cool, lets rename it to params.

caridy commented on lib/middleware.js:
```diff
@@ -100,23 +100,46 @@ function middleware() {
 
         store = mojito.store;
         appConfig = store.getStaticAppConfig();
+        midConfig = {
```
I will vote for `libmojito.middleware()` as a middleware function. then within that function having an internal cache routine that triggers on the first request, to be able to pick `app.mojito`, and `app.mojito.store` to collect any app level settings that is needed by the middleware to function.

imalberto replied:
@caridy That's the approach Mojito took for all its built-in middleware.

This specific implementation is to provide a level of BC for all other middleware that relies on the internal store, logger, etc. If not, that will require all of them to be updated.

In addition, we can document that going forward, the recommended way is to do as you suggested, and that this is mark a deprecated feature and will be removed in the future. /cc @zhouyaoji

drewfish replied:
What about just having mojito (or the user's app.js) just `app.set('store', mojito.store)` instead? Same approach for other midConfig fields as well.  Then each middleware can simply `req.app.get('store')`.

imalberto replied:
@drewfish I like this idea overall instead of just tagging onto `app.mojito` namespace! The only thing is that `"store"` is very generic, and I am thinking perhaps a more mojito specific NS should be used. e.g. `app.set('mojito store', mojito.store');`.

For this specific issue here, search team have quite a few mojito-\* extensions out there with middleware that relies on `midConfig` being passed to for initialization, and having to delay the initialization until that first request will basically break those. We can provide a level of BC but make sure we document those as _deprecated_ and _not recommented_ for future use, which I think is the least resistant path.

drewfish replied:
Yeah, I agree about the change where initialization occurs on the first request.

I agree that app.set('mojito store') is probably better.

As far as search middleware boot-up, remember that they are in control app.js which means that they can load there middleware however they want. (They don't necessarily need to let mojito do it.) Hmmm... the issue with what I'm describing is that we do want sugar for mojito to register the middleware that it knows about (the middleware that the RS found). If they wanted to instantiate (with stuff like midConfig) middleware by hand (not using mojito's sugar) then they still need to find out the list of middleware that mojito knows about, but this is just a RS.getResourceVersions() call so shouldn't be too hard.

caridy replied:
I like `mojito store` as a config. we are already doing something similar with express-yui, by using `app.set('locator', locatorInstance);`

imalberto replied:
I've updated `lib/mojito` so that now `app.js` can access the store if needed:

```
// app.js
...
store = app.get('mojito store'); // access to the resource store
versions = store.getResourceVersions(...);
...
```

We can update the built in middleware to use that new interface in a future PR.

caridy commented on lib/mojito.js:
```diff
@@ -8,29 +8,21 @@
+    mojito = app.mojito;
 
-    app.mojito.attachRoutes(); // load routes defined in `routes.json`
+    app.use(mojito.middleware());
```
isn't this suppose to happen in the app.js?

imalberto replied:
You are correct. I think that specific line though is part of the header file documentation showing basic usage for now to extend `app` with `mojito`.

caridy commented on lib/mojito.js:
```diff
@@ -251,28 +237,20 @@ Mojito.prototype._configureAppInstance = function (app, store, options, appConfi
+module.exports = {
+    // BEGIN DO NOT USE - for testing only
+    // @private
+    Mojito: Mojito,
```
why? normally, if you do the proper extend (following the principles described by @ericf) you should be able to get a handle on the original class by looking into the defined property.

imalberto replied:
Good point!

caridy commented:
it is looking good. small details here and there, but in general, we are pretty close to get this one merged.

lzhan commented on tests/func/serveronly/serveronlytest_descriptor.json:
```diff
@@ -550,6 +586,12 @@
                         {
+                            "controller": "locator",
+                            "params": {
+                                "value": "#output"
```
The element id for this test should be #my_header.
This will fix the travis failure on the serveronly app test.

imalberto replied:
Fixed.

Commit:
Merge remote-tracking branch 'alberto/search-next' into next-test

Commit:
fix an element id

imalberto committed:
Removed `Mojito` constructor function from the the list of exports. Tests fixes.

imalberto committed:
merged downstream from lzhan/next-test

drewfish commented on lib/mojito.js:
```diff
@@ -87,46 +91,33 @@ Creates a new instance of mojito and attach to `app`.
 
     if (!app.mojito) {
-        throw new Error('`app.mojito` was not initialized correctly.');
+        throw new Error('`app.mojito` was not initialized propertly.');
```
"propertly" ?

imalberto committed:
Updated examples/ and tests/func/applications/ "app.js"

imalberto committed:
Added `dispatch` and `tunnelMiddleware` to the `mojito` module exports.

imalberto committed:
- added libmojito.dispatcher as an alias for libmojito.dispatch
- `mojito` and `resource store` references are now available via `app.get('mojito')` and `app.get('mojito store')`
- removed invalid modules from `require` list for `serveronly` functional test

imalberto committed:
Added mock for `app.set`.

imalberto committed:
Fixed examples to use `libmojito.middleware()` vs `app.mojito.middleware()`

imalberto committed:
Updated HISTORY_next.md

imalberto committed:
Removed references to app.set('mojito store'). Libraries and fws should not be setting on the `app` namespace.

imalberto committed:
Updating more examples to remove references to `app.mojito.middleware()`.

caridy commented:
+1
